rtengine/ipwavelet.cc

```diff
@@ -3400,6 +3403,394 @@ void ImProcFunctions::calckoe (const float* WavCoeffs, float gradw, float tloww,
+                    inva5 = 1.f;
+                    inva6 = 0.8f;
+                    inva7 = 0.65f;
+                    inva8 = 0.5f;
+                    inva9 = 0.3f;
+                    inva10 = 0.2f;
+                }
+            } else if( pyrwav == 3) {//default
+                level_bl = 0;
+                level_hl = 0;
+                level_br = wavelet_lev - 2;//-1
+                level_hr = wavelet_lev;//0
+                if(!cmparams.wsmoothcie) {
+                    inva5 = 1.f; //1
+                    inva6 = 0.9f;//0.9
+                    inva7 = 0.7f;//0.7
+                    inva8 = 0.6f;//0.6
+                    inva9 = 0.4f;//0.4
+                    inva10 = 0.2f;//0.2
+                }
+            } else if( pyrwav == 4) {
+                level_bl = 0;
+                level_hl = 0;
+                level_br = wavelet_lev -1;//0
+                level_hr = wavelet_lev +1;//0
+                if(!cmparams.wsmoothcie) {
+                    inva5 = 0.9f;
+                    inva6 = 0.8f;
+                    inva7 = 0.6f;
+                    inva8 = 0.5f;
+                    inva9 = 0.3f;
+                    inva10 = 0.1f;
+                }
+            } else if( pyrwav == 5) {
+                level_bl = 0;
+                level_hl = 0;
+                level_br = wavelet_lev - 1;//-1
+                level_hr = wavelet_lev + 2;//+1  //be careful the preview must be big enough to see the changes
+                if(!cmparams.wsmoothcie) {
+                    inva5 = 0.8f;//0.85
+                    inva6 = 0.6f;//0.75
+                    inva7 = 0.5f;//0.55
+                    inva8 = 0.3f;//0.45
+                    inva9 = 0.2f;//0.3
+                    inva10 = 0.05f;//0.1
+                    
+                }//last choice not used for various reasons
+            } else if( pyrwav == 6) {//agresive - maximum - in this case LUT are minimal to avoid artifacts -be careful the preview must be big enough to see the changes
+                level_bl = 0;
+                level_hl = 0;
+                level_br = wavelet_lev ;//-1
+                level_hr = wavelet_lev + 2;//here maximum
+            }
+
+            //find possible max levels in function of windows preview size.
+            int minwin = rtengine::min(width, height);
+            int maxlevelspot = 9;//maximum possible
+
+            // adapt maximum level wavelet to size of crop
+            while ((1 << maxlevelspot) >= minwin && maxlevelspot  > 1) {
+                --maxlevelspot ;
+            }
+
+            int wavelet_level = rtengine::min(level_hr, maxlevelspot);
+            int maxlvl = wavelet_level;
+            maxlevpo = maxlvl;
+            //decomposition wavelet , we can change Daublen (moment wavelet) in Tab - Wavelet Levels with subsampling = 1
+            wavelet_decomposition *wdspot = new wavelet_decomposition(lab->L[0], width, height, maxlvl, 1, skip, numThreads, DaubLen);
```

Comment:
```suggestion
            std::unique_ptr<wavelet_decomposition> wdspot = std::unique_ptr<wavelet_decomposition>(new wavelet_decomposition(lab->L[0], width, height, maxlvl, 1, skip, numThreads, DaubLen));
```
Use a unique pointer to avoid accidentally leaking memory.

Reply:
OK, no problem